### PR TITLE
ENC Samples

### PR DESCRIPTION
**Pull request by nCastle1**
Demonstrates: adding an ENC Exchange Set to a map, selecting ENC features, and setting ENC display settings.

### CODE REVIEW

**Comment by nCastle1** [comment]
@dkemlage Can you please take a look?

I broke the pattern of doing only one sample per PR because a) the 3 samples share the majority of their code and b) the samples are all in the same category, and resolving the merge conflicts 3 times might be more work than writing the sample was ;)

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/AddEncExchangeSet/AddEncExchangeSet.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+            // Get the path to the ENC Exchange Set
+            string encPath = await GetEncPath();
+
+            // Create the Exchange Set
+            // Note: this constructor takes an array of paths because so that update sets can be loaded alongside base data
+            EncExchangeSet _encExchangeSet = new EncExchangeSet(new string[] { encPath });
+
+            // Wait for the exchange set to load
+            await _encExchangeSet.LoadAsync();
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/AddEncExchangeSet/AddEncExchangeSet.xaml.cs]: Rename the variable to 'myEncExchangeSet' as this is not a global/member variable.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/AddEncExchangeSet/AddEncExchangeSet.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+            // Store a list of data set extent's - will be used to zoom the mapview to the full extent of the Exchange Set
+            List<Envelope> dataSetExtents = new List<Envelope>();
+
+            // Add each data set as a layer
+            foreach (EncDataSet _encDataSet in _encExchangeSet.DataSets)
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/AddEncExchangeSet/AddEncExchangeSet.xaml.cs]: Rename the variable to 'myEncDataSet' as this is not a global/member variable.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/AddEncExchangeSet/AddEncExchangeSet.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            {
+                // Create the cell and layer
+                EncLayer _encLayer = new EncLayer(new EncCell(_encDataSet));
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/AddEncExchangeSet/AddEncExchangeSet.xaml.cs]: Rename the variable to 'myEncLayer' as this is not a global/member variable.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/AddEncExchangeSet/AddEncExchangeSet.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+        private async Task<String> GetEncPath()
+        {
+            #region offlinedata
+
+            // The data manager provides a method to get the folder
+            string folder = DataManager.GetDataFolder();
+
+            // Get the full path
+            string filepath = Path.Combine(folder, "SampleData", "AddEncExchangeSet", "ExchangeSet", "ENC_ROOT", "CATALOG.031");
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/AddEncExchangeSet/AddEncExchangeSet.xaml.cs]: Can we add some comments on what the parameters of the Path.Combine mean? Perhaps something like:
// The resulting path will look something like:
// "<InstallLocationOnDisk>\SampleData\AddEncExchangceSet\ExchangeSet\ENC_ROOT\Catalog.031"

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+using Esri.ArcGISRuntime.Geometry;
+using Esri.ArcGISRuntime.Hydrography;
+using Esri.ArcGISRuntime.Mapping;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using System.Windows;
+
+namespace ArcGISRuntime.WPF.Samples.EncDisplaySettings
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs]: Unfortunately we are going to have to rename this sample. We have problems when the name of a sample is the same an and ESRI type, which in this case also happens to be: Esri.ArcGISRuntime.Hydrography.EncDisplaySettings. I would recommend: Change ENC display settings (name on disk --> ChangeEncDisplaySettings).

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+            // Initialize the map with an oceans basemap
+            MyMapView.Map = new Map(Basemap.CreateOceans());
+
+            // Get the path to the ENC Exchange Set
+            string encPath = await GetEncPath();
+
+            // Create the Exchange Set
+            // Note: this constructor takes an array of paths because so that update sets can be loaded alongside base data
+            EncExchangeSet _encExchangeSet = new EncExchangeSet(new string[] { encPath });
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs]: Rename the variable to 'myEncExchangeSet' as this is not a global/member variable.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+            // Wait for the exchange set to load
+            await _encExchangeSet.LoadAsync();
+
+            // Store a list of data set extent's - will be used to zoom the mapview to the full extent of the Exchange Set
+            List<Envelope> dataSetExtents = new List<Envelope>();
+
+            // Add each data set as a layer
+            foreach (EncDataSet _encDataSet in _encExchangeSet.DataSets)
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs]: Rename the variable to 'myEncDataSet' as this is not a global/member variable.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            {
+                // Create the cell and layer
+                EncLayer _encLayer = new EncLayer(new EncCell(_encDataSet));
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs]: Rename the variable to 'myEncLayer' as this is not a global/member variable.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+        private async Task<String> GetEncPath()
+        {
+            #region offlinedata
+
+            // The data manager provides a method to get the folder
+            string folder = DataManager.GetDataFolder();
+
+            // Get the full path
+            string filepath = Path.Combine(folder, "SampleData", "AddEncExchangeSet", "ExchangeSet", "ENC_ROOT", "CATALOG.031");
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs]: Can we add some comments on what the parameters of the Path.Combine mean? Perhaps something like:
// The resulting path will look something like:
// "\SampleData\AddEncExchangceSet\ExchangeSet\ENC_ROOT\Catalog.031"

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+            return filepath;
+
+            #endregion offlinedata
+        }
+
+        private void UpdateDisplaySettings()
+        {
+            // Apply color scheme
+            if ((bool)radDay.IsChecked) { EncEnvironmentSettings.Default.EncDisplaySettings.MarinerSettings.ColorScheme = EncColorScheme.Day; }
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/Hydrography/EncDisplaySettings/EncDisplaySettings.xaml.cs]: We really need to break out all of these one-liner chained statements into smaller atomic statements and provide comments on what is going on using our API. I would suggest something like:
```
            if ((bool)radDay.IsChecked)
            {
                // Get the default ENC environment settings
                EncEnvironmentSettings myEncSettings_Default = EncEnvironmentSettings.Default;

                // Get the ENC display settings
                EncDisplaySettings myEncDisplaySettings = myEncSettings_Default.EncDisplaySettings;

                // Get the S52 display mariner settings
                EncMarinerSettings myEncMarinerSettings = myEncDisplaySettings.MarinerSettings;

                // Set the S52 mariner settings to the ENC color scheme of day
                myEncMarinerSettings.ColorScheme = EncColorScheme.Day;
            }
```

**Pull request merged**